File: WEEK2/bj7_11724.py
import sys

sys.setrecursionlimit(10**6)


def dfs(node):
    visited[node] = True
    for neighbor in adj_list[node]:
        if not visited[neighbor]:
            dfs(neighbor)


n, m = map(int, sys.stdin.readline().split())

adj_list = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)

for _ in range(m):
    u, v = map(int, sys.stdin.readline().split())
    adj_list[u].append(v)
    adj_list[v].append(u)

component_count = 0
for i in range(1, n + 1):
    if not visited[i]:
        component_count += 1
        dfs(i)

print(component_count)

File: WEEK2/bj7_11724.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const edges = input.slice(1);

const adjList = Array.from({ length: n + 1 }, () => []);
const visited = new Array(n + 1).fill(false);

for (let i = 0; i < m; i++) {
    const [u, v] = edges[i].split(' ').map(Number);
    adjList[u].push(v);
    adjList[v].push(u);
}

function dfs(node) {
    visited[node] = true;
    for (const neighbor of adjList[node]) {
        if (!visited[neighbor]) {
            dfs(neighbor);
        }
    }
}

let componentCount = 0;
for (let i = 1; i <= n; i++) {
    if (!visited[i]) {
        componentCount++;
        dfs(i);
    }
}

console.log(componentCount);
